### ISSUE TITLE
MacOS system ruby (universal.x86_64-darwin19) chooses to install arm64-darwin native gem

### ISSUE THREAD

**Issue by flavorjones** [rubygems/rubygems#4234]
:wave: I'm a maintainer of Nokogiri, and we recently started shipping native gems for darwin:

- https://rubygems.org/gems/nokogiri/versions/1.11.0-x86_64-darwin
- https://rubygems.org/gems/nokogiri/versions/1.11.0-arm64-darwin

We received a bug report yesterday (https://github.com/sparklemotion/nokogiri/issues/2165) from @rgoldberg that I'm not sure how to address, and I'm hoping someone from the RubyGems team can help me understand what's going on.

**Please describe what you're seeing**

On an Intel-based mac running Catalina (10.15.7), using the system-installed ruby (`/usr/bin/ruby`) and gem (`/usr/bin/gem`) which reports:

```
RubyGems Environment:
  - RUBYGEMS VERSION: 3.0.3
  - RUBY VERSION: 2.6.3 (2019-04-16 patchlevel 62) [universal.x86_64-darwin19]
  - INSTALLATION DIRECTORY: /Library/Ruby/Gems/2.6.0
  - USER INSTALLATION DIRECTORY: /Users/flavorjones/.gem/ruby/2.6.0
  - RUBY EXECUTABLE: /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
  - GIT EXECUTABLE: /Users/flavorjones/.nix-profile/bin/git
  - EXECUTABLE DIRECTORY: /usr/local/bin
  - SPEC CACHE DIRECTORY: /Users/flavorjones/.gem/specs
  - SYSTEM CONFIGURATION DIRECTORY: /Library/Ruby/Site
  - RUBYGEMS PLATFORMS:
    - ruby
    - universal-darwin-19
  - GEM PATHS:
     - /Library/Ruby/Gems/2.6.0
     - /Users/flavorjones/.gem/ruby/2.6.0
     - /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/gems/2.6.0
  - GEM CONFIGURATION:
     - :update_sources => true
     - :verbose => true
     - :backtrace => false
     - :bulk_threshold => 1000
     - "gem" => "--no-ri --no-rdoc --no-document"
  - REMOTE SOURCES:
     - https://rubygems.org/
  - SHELL PATH:
     - <redacted>
     - /Applications/Emacs.app//Contents/MacOS/bin-x86_64-10_10
```

Extracting the relevant bits:
  - RUBY EXECUTABLE: /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
  - RUBY VERSION: 2.6.3 (2019-04-16 patchlevel 62) [universal.x86_64-darwin19]
  - RUBYGEMS PLATFORMS:
    - ruby
    - universal-darwin-19

When I run `gem install nokogiri`, RubyGems -- sometimes! -- chooses to install the `arm64-darwin` platform of the gem, instead of the `x86_64-darwin` platform of the gem. Yesterday I was able to reproduce this at will, and this morning I cannot.

When the "wrong" native gem does get installed, users see an error message like this when they try to use Nokogiri:

> /Library/Ruby/Site/2.6.0/rubygems/core_ext/kernel_require.rb:54:in `require': dlopen(/Library/Ruby/Gems/2.6.0/gems/nokogiri-1.11.0-arm64-darwin/lib/nokogiri/2.6/nokogiri.bundle, 0x0009): could not use '/Library/Ruby/Gems/2.6.0/gems/nokogiri-1.11.0-arm64-darwin/lib/nokogiri/2.6/nokogiri.bundle' because it is not a compatible arch - /Library/Ruby/Gems/2.6.0/gems/nokogiri-1.11.0-arm64-darwin/lib/nokogiri/2.6/nokogiri.bundle (LoadError)

**Expected behavior**

I would expect rubygems to always prefer the `x86_64-darwin` platform gem on a system that is Intel-based; and to always prefer the `arm64-darwin` platform gem on an ARM-based system.

**Additional Info**

Yesterday I was able to reproduce this every time. This morning, I can't get it install anything except the default `ruby` platform (that is, I can't get it to install native platforms at all). Which makes me wonder whether there's some non-deterministic thing happening, perhaps with the order in which the API returns potential matches? Just a guess.

There is more information from the original reporter at https://github.com/sparklemotion/nokogiri/issues/2165 if it's helpful.

---

I will abide by the [code of conduct](https://github.com/rubygems/rubygems/blob/master/CODE_OF_CONDUCT.md).

**Comment by rgoldberg** [comment]
I was the reporter of sparklemotion/nokogiri#2165.

I'm running Big Sur 11.1 on an Intel Mac Mini 2014, which seems to be Darwin 20.2.0.

So the issue affects both Catalina & Big Sur.

**Comment by deivid-rodriguez** [comment]
@flavorjones Something I don't understand is that I believe your local platform should be reported as `universal.x86_64-darwin-19`, not as `universal-darwin-19`. And samething for @rgoldberg. Not sure why it's not reported like that... :grimacing:.

What's the result of `Gem::Platform.local` and `RbConfig::CONFIG["arch"]`?

**Comment by flavorjones** [comment]
Here's what it is on my system:

```
~ $ ruby -v
ruby 2.6.3p62 (2019-04-16 revision 67580) [universal.x86_64-darwin19]

~ $ irb

WARNING: This version of ruby is included in macOS for compatibility with legacy software.
In future versions of macOS the ruby runtime will not be available by
default, and may require you to install an additional package.

irb(main):001:0> Gem::Platform.local.to_s
=> "universal-darwin-19"
irb(main):002:0> Gem::Platform.local
=> #<Gem::Platform:0x00007fc0b70bc708 @cpu="universal", @os="darwin", @version="19" @cpu="universal", @os="darwin", @version="19">
irb(main):003:0> RbConfig::CONFIG["arch"]
=> "universal-darwin19"
```

**Comment by deivid-rodriguez** [comment]
Well that explains things, since `rubygems` uses information from `RbConfig::CONFIG["arch"]` to setup its local platform, and there's nothing about intel vs arm in there. I guess we should use `RUBY_PLATFORM` instead?

**Comment by flavorjones** [comment]
¯\\\_(ツ)\_/¯  I'm not sure what the right thing to do is. I can say that `RUBY_PLATFORM` does seem like a better choice in this particular instance:

```
irb(main):004:0> RUBY_PLATFORM
=> "universal.x86_64-darwin19"
```

and that this value matches the value of `RbConfig::CONFIG['arch']` in all my other ruby installations across Linux and Darwin.

**Comment by flavorjones** [comment]
Although, if we used `RUBY_PLATFORM` then neither of the native gems would match for this ruby (and maybe that's OK):

```
irb(main):004:0> RUBY_PLATFORM
=> "universal.x86_64-darwin19"
irb(main):005:0> Gem::Platform.new(RUBY_PLATFORM)
=> #<Gem::Platform:0x00007fc0a70c1038 @cpu="universal.x86_64", @os="darwin", @version="19" @cpu="universal.x86_64", @os="darwin", @version="19">
irb(main):006:0> Gem::Platform.new(RUBY_PLATFORM) =~ Gem::Platform.new("x86_64-darwin")
=> false
irb(main):007:0> Gem::Platform.new(RUBY_PLATFORM) =~ Gem::Platform.new("arm64-darwin")
=> false
```

**Comment by deivid-rodriguez** [comment]
In principle I would say `universal.x86_64-darwin19` should match `x86_64-darwin` and not match `arm64-darwin`, so we should fix platform matching for this case too. But I'm not fully sure.

@nobu Can you help us?

**Comment by hsbt** [comment]
System Ruby of Apple applied the custom patch, and the ruby core team couldn't handle it.

**Comment by deivid-rodriguez** [comment]
Thanks @indirect, that's helpful.

So that means we don't want https://github.com/rubygems/rubygems/pull/4238/commits/27d34081f5f30a5e25dda48b9c532caa16a02ae4, since it's incorrect, but instead what @flavorjones demostraded in https://github.com/rubygems/rubygems/issues/4234#issuecomment-754840599 is the correct behaviour where we will fallback to the standard ruby variant since nothing else will match in this case.

I was reading about [a related discussion for python](https://discuss.python.org/t/apple-silicon-and-packaging/4516/18) and got the impression though that fat binary gems made for x86_64 should be compatible with this new "dual platform".

**Comment by indirect** [comment]
I believe that's correct—my understanding is that Apple's term "universal" means that the fat binary contains both an x86 and an ARM compiled binary, and can run natively on either architecture. That said, it might be up to every individual gem to adjust their extconf compilation steps to support universal builds vs just building a binary for only the arch that is running the `gem install` 😬

**Comment by indirect** [comment]
(and I just realized that this is probably even more complicated than what I just said here... I think compiled gem extensions that link against other installed .sos would also need those other .so files to also be universal, but Homebrew (for example) does not provide universal builds of things like mysql. So maybe universal gems is not possible in a practical way. :/

Does that mean RubyGems should pretend to be whichever binary is currently running, and if you change architectures everything will just break? that does not seem ideal either, but I don't know what other options there are if you are building gems that link against non-universal libraries.

**Comment by rgoldberg** [comment]
Is `universal.x86_64-darwin` output when the file is universal but the current CPU is x86_64?

If so, I assume the problem is that the same Ruby distribution could be shared between multiple computers (or used as an install template for both x86 & arm machines), so the same universal file (or a copy thereof) could also be run by a arm64 CPU, which would presumably output `universal.arm64-Darwin`.

If this is the issue, should all binary gems, including nokogiri, only distribute universal binaries? If so, on Catalina and earlier, it should be fine to install x86_64-only binaries, whereas arm64 (and possibly universal, if the version doesn’t support them) binaries should not be installed. On Big Sur & newer, maybe gem should reject installing non-universal binaries, or output an overrideable error or warning when installing a platform-specific binary.

**Comment by hsbt** [comment]
`universal` means the different feature in M1 chip. In M1 environment, `universal` will pick `x86_64` arch with Rosetta2 or `arch -x86_64` command. But It will pick `arm64` in natively. 

I'm not sure what the correct arch in M1 chip yet. So, We need to investigate it.

**Comment by indirect** [comment]
> Is universal.x86_64-darwin output when the file is universal but the current CPU is x86_64?

No. As far as I know, is output when the file is universal and the current CPU is not considered—it could be x86, it could be ARM, and it could be ARM transpiling the current binary to from x86 via Rosetta.

Further complicating things, Apple Silicon machines like the M1 can run _both_ the x86 and ARM binaries, at any point. Universal gems would need to also contain both x86 and ARM binaries to work correctly on both CPU architectures on the same machine.

**Comment by rgoldberg** [comment]
> As far as I know, [`universal.x86_64-darwin`] is output when the file is universal and the current CPU is not considered

Why would Apple output `universal.x86_64` when the current CPU isn't considered? Why wouldn't they just output `universal` when they don't know the architecture?

I can't think of any reasonable reason for them to do that. I'm not saying they don't do that, but if anyone knows a good reason why they'd do that (other than Apple having terrible software policies), please let me know so I can learn their logic.

**Comment by rgoldberg** [comment]
I don't know how `gem` interacts with gems, but is there any way you could force gems to provide 100% universal binaries for Big Sur & later?

Maybe at first just warn if any non-universal files are created. Then later output an error that can be overridden with an option. Then later output a non-overrideable error.

Why would someone not provide universal binaries, other than laziness or dependencies not yet having universal binaries?

**Comment by rgoldberg** [comment]
I guess there are a few issues here:

1. Whether to allow non-universal binaries for Big Sur & later

If no:

2. How to enforce universal-only binaries

If yes:

2. How to automatically select the correct architecture if non-universal are allowed

3. Whether (& how) to allow allow manual override of architecture selection (x86_64, arm64, or universal)

**Comment by hsbt** [comment]
> Why would someone not provide universal binaries, other than laziness or dependencies not yet having universal binaries?

Because Apple didn't disclose its build instructions for RubyGems.

**Comment by rgoldberg** [comment]
@eregon What are the hard parts about producing a universal binary?

I develop for the JVM, so I haven't looked into it, but wouldn't it just be ensuring that a project's dependencies are available as universal binaries, then changing a few build options?

Is it much more involved than that?

**Comment by eregon** [comment]
> @eregon What are the complexities of producing a universal binary?

For starters, one would need an M1 mac, and I'd think most CIs or many devs don't have that.

Then it seems to need extra options and even extra commands and steps (e.g., https://developer.apple.com/documentation/xcode/building_a_universal_macos_binary seems quite involved).
From looking at a few extconf.rb, the darwin platform already needs extra flags/hacks, and this would only make it worse and harder to maintain.

I believe it's general knowledge that the system Ruby on macOS is basically unusable and not recommended (notably, can't install gems without `sudo`, and with `sudo` it's asking for troubles, this universal thing, etc).

**Comment by rgoldberg** [comment]
@eregon From what I've read, you can build universal binaries on either Apple Silicon or on Intel.

System Ruby versus other Ruby installs has nothing to do with building a universal binary.

At least some of the extra steps in the article to which you linked would also have to be done to handle building separate binaries for the architectures, not just to build universal binaries. So those are extra complexities of universal binaries, they’re complexities of having more than one architecture.

**Comment by flavorjones** [comment]
@rgoldberg I appreciate your enthusiasm for this topic, but I want you to understand that as a gem maintainer I have a toolchain (rake-compiler) that can build architecture-specific native gems, and I do not have a toolchain for building universal binaries. If you're interested in helping extend rake-compiler to support this workflow, please do -- but it is off-topic here and I'd like to suggest we stick to solving the problem at hand in this thread.

**Comment by drbrain** [comment]
Here are the [platform matching rules](https://github.com/rubygems/rubygems/blob/41142062b26f82599d074a519431b072e0e277db/lib/rubygems/commands/help_command.rb#L237-L243)

Since a given ruby installation supports exactly one CPU architecture at a time, I think having a `universal` included `Gem.platforms` doesn't make sense, it should only list the CPU architecture running.  For a `RUBY_PLATFORM` of `"universal.x86_64-darwin19"` RubyGems should disregard `universal` and use only `x86_64`.

Where `universal` CPU architecture does make sense is in a gem, where the author has compiled shared objects for multiple versions and put them all in the same gem, along with some code to require the correct architecture version based on `Gem.platforms` or equivalent. Presumably the author will take care to cover the reasonable CPU architectures for the OS they're targeting.

#4238 looks good to me as it drops `universal.` and from my reading leaves the actual CPU in Gem::Platform.local.

**Comment by Bo98** [comment]
I don't think #5888 fixes this. #4238 should however (though arm64e should be mapped to arm64).

**Comment by deivid-rodriguez** [comment]
Thanks for the heads up, can you share steps to reproduce the problem?

**Comment by deivid-rodriguez** [comment]
Can you actually upgrade RubyGems on MacOS system install. As far as I'm aware, MacOS tries very hard to avoid users writing into privileged locations like where system Ruby is installed.

If this is the case, it seems pointless to try fix this ourselves because the fix will never reach MacOS users, since I believe MacOS developers have discontinued their version of system Ruby. Also, this is actually fixed in core Ruby, so if MacOS developers wanted to fix this they would simply need to stop patching Ruby.

**Comment by deivid-rodriguez** [comment]
> I agree with this somewhat - though the issue is that macOS Ruby is too old rather than anything to do with patches (macOS Ruby is 2.6, but the core Ruby fix was made in 3.1).

Both actually because if I recall correctly they were precisely patching the code that got fixed. So if they upgraded the Ruby they ship, they'd need to not erase the fix 😅.

> Not exactly great but works. I don't know if there's any scope or interest at all in Bundler doing something better here, maybe not. But the above what I'm rolling with at the moment.

So what would work for you would be that Bundler actually fixed this I guess, right? We have a `lib/bundler/rubygems_ext.rb` in Bundler where we apply patches to RubyGems to get a consistent Bundler behavior across all RubyGems versions. Maybe that's a good place to fix this?

**Comment by Bo98** [comment]
> Both actually because if I recall correctly they were precisely patching the code that got fixed. So if they upgraded the Ruby they ship, they'd need to not erase the fix 😅.

Ah I see what one you're referring too now. I'll ask about that but it originated well before arm64 was a thing - it was patched out back in Ruby 2.0 at least. I guess no one noticed given what core Ruby was doing until recently didn't exctly work anyway unless you specified some env variables, which seems volatile.

> So what would work for you would be that Bundler actually fixed this I guess, right? We have a `lib/bundler/rubygems_ext.rb` in Bundler where we apply patches to RubyGems to get a consistent Bundler behavior across all RubyGems versions. Maybe that's a good place to fix this?

I think so yeah.

**Comment by Bo98** [comment]
Ok thinking into it more why, it actually does make a lot of sense why things are the way they are.

Apple ship universal binaries for everything. The same install is designed to work on _both_ x86_64 and arm64. So for example we have:

```
$ file /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/universal-darwin21/zlib.bundle
/System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/universal-darwin21/zlib.bundle: Mach-O universal binary with 2 architectures: [x86_64:Mach-O 64-bit bundle x86_64
- Mach-O 64-bit bundle x86_64] [arm64e:Mach-O 64-bit bundle arm64e
- Mach-O 64-bit bundle arm64e]
/System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/universal-darwin21/zlib.bundle (for architecture x86_64):	Mach-O 64-bit bundle x86_64
/System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/universal-darwin21/zlib.bundle (for architecture arm64e):	Mach-O 64-bit bundle arm64e
``` 

So the extension directory `universal-darwin21` makes perfect sense because of that.

For user installs, it does behave a little differently:

```
$ file /Library/Ruby/Gems/2.6.0/extensions/universal-darwin-21/2.6.0/nokogiri-1.13.8/nokogiri/nokogiri.bundle 
/Library/Ruby/Gems/2.6.0/extensions/universal-darwin-21/2.6.0/nokogiri-1.13.8/nokogiri/nokogiri.bundle: Mach-O 64-bit bundle x86_64
```

This partially my fault. Up until arm64 support was added, this did used to also be built universally. However when arm64 was added, `ARCH_FLAGS` would contain `-arch arm64e -arch x86_64`. `arch64e` is some in-development ABI that only Apple are allowed to use at the moment, so it caused errors for user installs. When I told them about it, the fix was to just blank arch flags and let the compiler default kick in. This fixed the issues I was having. The better solution in hindsight was probably `s/arm64e/arm64`  (`arm64` and `arm64e` binaries can talk to each other fine).

But that's all minor points really, given no one has really complained about that (switching between archs doesn't happen too often) and it can be overridden by `ENV["ARCHFLAGS"]` anyway.

So the problem really is only when it comes to gems that have been published prebuilt as separate archs. So something that affects this selection like `Gem.platforms` makes sense over something like `Gem::Platform.local` that will change everything including the extensions directory, which Apple would likely continue to revert.

**Comment by deivid-rodriguez** [comment]
Feel free to send us a patch to make Bundler patch `Gem.platforms` so that it gives you what you expect in this particular case!

**Comment by Bo98** [comment]
I'll have a look.

**Comment by rbrooks** [comment]
Doesn't the `-19` suffix lock it down to Major OS ver (Catalina)? Everyone on your team will probably be experiencing unpredictable `bundle installs` on non-Catlina OSes, on Gems with native Binaries. I had to do this kind of thing in Gemfile.lock. Omit the numeric suffix:

```
PLATFORMS
  arm64-darwin
  ruby
  x86_64-darwin
  x86_64-linux
```

Of course, do it via the `platform` commands; not editing Lock directly:

```
bin/bundle lock --remove-platform arm64-darwin-21
bin/bundle lock --add-platform arm64-darwin
```

**Comment by Bo98** [comment]
Yes that's correct, and I too recommend omitting the numeric suffix in Bundler lockfiles, though that's a completely separate issue to what was discussed here.

The issue here is that the `arm64-darwin` version was _always_ used on a `universal-darwin` build of Ruby, even when running on x86_64. I made a fix for that which has been merged recently for Bundler (on `master`).

For RubyGems (`gem install`), there's not much point changing much given it's fixed in Ruby 3.1 and RubyGems doesn't support Apple's Ruby anymore so a backport for that would require more than just a single fix.
